lms/services/rsa_key.py

```diff
@@ -17,6 +18,51 @@ def __init__(self, db, aes_service):
         self._db = db
         self._aes_service = aes_service
 
+    def rotate(
+        self,
+        target_keys: int,
+        max_age=timedelta(days=7),
+        max_expired_age=timedelta(days=14),
+    ):
+        """
+        Rotate RSA Keys keeping `target_keys` active keys at all times.
+
+        Expires active keys older than `max_age`
+        and deletes expired ones older than `max_expired_age`.
+
+        Keys have a live-cycle of:
+            active -> expired -> deleted
+
+        In flight request we signed with an active key will still be
+        verifiable as the public key endpoint query (get_all_public_jwks)
+        includes expired keys.
+
+        This method is meant to be called periodically by a celery task.
+        """
+        now = datetime.now()
+        new_keys = target_keys
+
+        for key in self._db.query(RSAKey).all():
+            key_age = now - key.created
+            # Expire valid keys after max_age
+            if not key.expired and key_age >= max_age:
+                key.expired = True
+            # Delete expired keys that have been around for `max_expired_age`
+            if key.expired and key_age >= max_expired_age:
```

Comment:
Comparing against the same date in both cases but only deleting ones that are already expired.

Reply:
Yeah. I think this is a best of both!